components/net/bluetooth_thread.rs

```diff
@@ -0,0 +1,371 @@
+                }
+                BluetoothMethodMsg::GetPrimaryService(device_id, sender) => {
+                    self.get_primary_service(device_id, sender)
+                }
+                BluetoothMethodMsg::GetCharacteristic(service_id, sender) => {
+                    self.get_characteristic(service_id, sender)
+                }
+                BluetoothMethodMsg::GetDescriptor(characteristic_id, sender) => {
+                    self.get_descriptor(characteristic_id, sender)
+                }
+                BluetoothMethodMsg::ReadValue(id, sender) => {
+                    self.read_value(id, sender)
+                }
+                BluetoothMethodMsg::WriteValue(id, value, sender) => {
+                    self.write_value(id, value, sender)
+                }
+                BluetoothMethodMsg::Exit => {
+                    break
+                }
+            }
+        }
+    }
+
+    fn request_device(&mut self, sender: IpcSender<BluetoothObjectMsg>) {
+        if !self.adapter.is_initialized() {
+            sender.send(BluetoothObjectMsg::Error { error: String::from("No adapter found") }).unwrap();
+            return;
+        }
+        let devices = self.adapter.get_devices(true);
+        if devices.is_empty() {
+            sender.send(BluetoothObjectMsg::Error { error: String::from("No device found") }).unwrap();
+            return;
+        }
+
+        for device in devices.clone() {
+            self.cached_devices.insert(device.get_address(), device.clone());
+        }
+
+        let device = devices[0].clone();
+
+        let message = BluetoothObjectMsg::BluetoothDevice {
+            id: device.get_address(),
+            name: device.get_name(),
+            device_class: device.get_class(),
+            vendor_id_source: device.get_vendor_id_source(),
+            vendor_id: device.get_vendor_id(),
+            product_id: device.get_product_id(),
+            product_version: device.get_device_id(),
+            appearance: device.get_appearance(),
+            tx_power: device.get_tx_power() as i8,
+            rssi: device.get_rssi() as i8
+        };
+        sender.send(message).unwrap();
+    }
+
+    pub fn gatt_server_connect(&mut self, device_id: String, sender: IpcSender<BluetoothObjectMsg>) {
+        if !self.adapter.is_initialized() {
+            sender.send(BluetoothObjectMsg::Error { error: String::from("No adapter found") }).unwrap();
+            return;
+        }
+
+        let device = match self.adapter.get_device(device_id, true) {
+            Some(d) => d,
+            None => {
+                sender.send(BluetoothObjectMsg::Error { error: String::from("No device found") }).unwrap();
+                return;
+            }
+        };
+
+        //TODO check result
+        device.connect();
+        println!("Connected: {:?}", device.is_connected());
+        let message = BluetoothObjectMsg::BluetoothServer {
```

Comment:
Why don't you send the message directly? Like `sender.send(BluetoothObjectMsg::BluetoothServer...`